File: backend/functions/services/floorplan/floorplan.controller.js
//Floor plan controller handles the operations of the floor plan service with business logic and CRUD operations

const Room = require("../../models/room.model");
const uuid = require("uuid"); // npm install uuid

let database; //this variable holds the database
let reportingDatabase;

//This function allows dependency injections for production or mock databases
exports.setDatabase = async(db) => {
  database = db;
}

exports.setReportingDatabase = async (db) => {
  reportingDatabase = db;
}

/**
 * This function creates a specified floor plan via an HTTP CREATE request.
 * @param req The request object must exist and have the correct fields. It will be denied if not.
 * The request object should contain the following:
 *   numFloors: num
 *   adminId: string
 *   companyId: string
 * @param res The response object is sent back to the requester, containing the status code and a message.
 * @returns res - HTTP status indicating whether the request was successful or not.
 */
exports.createFloorPlan = async (req, res) => {
  let fieldErrors = [];

  if (req == null) {
    fieldErrors.push({field: null, message: 'Request object may not be null'});
  }

  //Look into express.js middleware so that these lines are not necessary
  let reqJson;
  try {
    reqJson = JSON.parse(req.body);
  } catch (e) {
    reqJson = req.body;
  }
  console.log(reqJson);
  //////////////////////////////////////////////////////////////////////

  if (reqJson.numFloors == null || reqJson.numFloors === '') {
    fieldErrors.push({field: 'numFloors', message: 'Number of floors may not be empty'});
  }

  if (reqJson.adminId == null || reqJson.adminId === '') {
    fieldErrors.push({field: 'adminId', message: 'Admin ID may not be empty'});
  }

  if (reqJson.companyId == null || reqJson.companyId === '') {
    fieldErrors.push({field: 'companyId', message: 'Company ID may not be empty'});
  }

  if (fieldErrors.length > 0) {
    console.log(fieldErrors);
    return res.status(400).send({
      message: '400 Bad Request: Incorrect fields',
      errors: fieldErrors
    });
  }

  try {
  let floorplanNumber = "FLP-" + uuid.v4();
    
  let floorplanData = {
    floorplanNumber: floorplanNumber,
    numFloors: reqJson.numFloors,
    adminId: reqJson.adminId,
    companyId: reqJson.companyId
  }
  await database.createFloorPlan(floorplanNumber,floorplanData);
  for (let index = 0; index < reqJson.numFloors; index++) {
  let floorNumber = "FLR-" + uuid.v4();
  let floorData = {
    floorNumber: floorNumber,
    numRooms: 0,
    currentCapacity: 0,
    maxCapacity: 0,
    floorplanNumber: floorplanNumber,
    adminId: reqJson.adminId,
    companyId:reqJson.companyId
  }
  await database.createFloor(floorNumber,floorData);
  console.log("Floor with floorNumber : "+floorNumber+" succesfully created under floorplan : "+floorplanNumber);

  }
  console.log("Floorplan with floorplanNumber : "+floorplanNumber+" succesfully created");

  // company-data summary
  let companyData = await reportingDatabase.getCompanyData(reqJson.companyId);
  await reportingDatabase.addNumberOfFloorplansCompanyData(reqJson.companyId, companyData.numberOfFloorplans);
  for (let i = 0; i < reqJson.numFloors; i++) {
    await reportingDatabase.addNumberOfFloorsCompanyData(reqJson.companyId, parseInt(companyData.numberOfFloors) + i);
  }

  return res.status(200).send({
  message: 'floorplan successfully created',
  data: reqJson
  });
  } catch (error) {
  console.log(error);
  return res.status(500).send(error);
  }

  };


  //This function adds a single floor to a specific floorplan
  exports.createFloor = async (req, res) => {
  try {
    let floorNumber = "FLR-" + uuid.v4();

    let reqJson = JSON.parse(req.body);
    console.log(reqJson);
    
    let floorData = {
      floorNumber: floorNumber,
      numRooms: 0,
      currentCapacity: 0,
      maxCapacity: 0,
      floorplanNumber: reqJson.floorplanNumber,
      adminId: reqJson.adminId,
      companyId:reqJson.companyId
    }

    let floorplan = await database.getFloorPlan(reqJson.floorplanNumber);
    await database.addFloor(reqJson.floorplanNumber,floorplan.numFloors);
    await database.createFloor(floorNumber,floorData);
    console.log("Floor with floorNumber : "+floorNumber+" succesfully created under floorplan : "+reqJson.floorplanNumber);
    
    // company-data summary
    let companyData = await reportingDatabase.getCompanyData(reqJson.companyId);
    await reportingDatabase.addNumberOfFloorsCompanyData(reqJson.companyId, companyData.numberOfFloors);

    return res.status(200).send({
      message: 'floor successfully created',
      data: floorData
    });
  } catch (error) {
    console.log(error);
    return res.status(500).send(error);
  }

};

//creating a room / adding a room to a floor 
// when creating a room , we need to initialize desks associated with that room as well
exports.createRoom = async (req, res) => {
try {
  let roomNumber = "RMN-" + uuid.v4();

  let reqJson = JSON.parse(req.body);
  console.log(reqJson);
  
  let room = new Room(reqJson.currentNumberRoomInFloor,roomNumber,"",reqJson.floorNumber,
      reqJson.roomArea,reqJson.deskArea, reqJson.numberDesks,reqJson.capacityPercentage);
  let roomData = {
    currentNumberRoomInFloor:room.currentNumberRoomInFloor,
    floorNumber:room.floorNumber,
    roomNumber:room.roomNumber,
    roomName:room.roomName,
    roomArea:room.roomArea, 
    capacityPercentage:room.capacityPercentage,
    numberDesks: room.numberDesks,
    occupiedDesks:room.occupiedDesks,
    currentCapacity:room.currentCapacity,
    deskArea:room.deskArea, 
    capacityOfPeopleForSixFtGrid:room.capacityOfPeopleForSixFtGrid,
    capacityOfPeopleForSixFtCircle:room.capacityOfPeopleForSixFtCircle
  }

  let floor = await database.getFloor(reqJson.floorNumber);

  await database.addRoom(reqJson.floorNumber,reqJson.currentNumberRoomInFloor);
  await database.createRoom(roomNumber,roomData);
  console.log("Room with roomNumber : "+roomNumber+" succesfully created under floor : "+reqJson.floorNumber);

  // company-data summary
  let companyData = await reportingDatabase.getCompanyData(floor.companyId);
  await reportingDatabase.addNumberOfRoomsCompanyData(floor.companyId, companyData.numberOfRooms);
  
  for (let index = 0; index < reqJson.numberDesks; index++) {
    let deskNumber = "DSK-" + uuid.v4();
    let deskData = {
      deskNumber: deskNumber,
      roomNumber: roomNumber,
      deskArea: reqJson.deskArea
    }
    await database.createDesk(deskNumber,deskData);
    console.log("Desk with deskNumber : "+deskNumber+" succesfully created under room : "+roomNumber);
    
  }
  
  return res.status(200).send({
    message: 'room successfully created',
    data: roomData
  });
} catch (error) {
  console.log(error);
  return res.status(500).send(error);
}

};

//This function returns all floorplans under a given company
//NB the spelling of companyId and companyId is very important for this function we query companyId
exports.viewFloorPlans = async (req, res) => {
try {
      let reqJson = JSON.parse(req.body);
      console.log(reqJson);
    let filteredList=[];
    let floorplans = await database.getFloorPlans();

    
      floorplans.forEach(obj => {
      if(obj.companyId===reqJson.companyId)
      {
        filteredList.push(obj);
      }
      else
      {
      }
    });
    
    return res.status(200).send({
      message: 'Successfully retrieved floorplans based on your company',
      data: filteredList
    });
} catch (error) {
    console.log(error);
    return res.status(500).send({
      message: error.message || "Some error occurred while fetching floorplans."
    });
}
};

//This function fetches all floors under a floorplan
exports.viewFloors = async (req, res) => {
try {
    let reqJson = JSON.parse(req.body);
    let filteredList=[];
    let floors = await database.getFloors();
    
    floors.forEach(obj => {
      if(obj.floorplanNumber===reqJson.floorplanNumber)
      {
        filteredList.push(obj);
      }
      else
      {

      }
    });
    return res.status(200).send({
      message: 'Successfully retrieved floors based on your floorplan',
      data: filteredList
    });
} catch (error) {
    console.log(error);
    return res.status(500).send({
      message: err.message || "Some error occurred while fetching floors."
    });
}
};

//we query rooms based on the floor they in using the floor number
exports.viewRooms = async (req, res) => {
try {
    let reqJson = JSON.parse(req.body);
    let filteredList=[];
    let rooms = await database.getRooms();
    
    rooms.forEach(obj => {
      if(obj.floorNumber===reqJson.floorNumber)
      {
        filteredList.push(obj);
      }
      else
      {

      }
    });
    return res.status(200).send({
      message: 'Successfully retrieved rooms based on your floor number',
      data: filteredList
    });
} catch (error) {
    console.log(error);
    return res.status(500).send({
      message: err.message || "Some error occurred while fetching rooms."
    });
}
};

  exports.createFloorPlanMock = async (floorplanNumber, data) => {
      return await database.createFloorPlan(floorplanNumber,data);

};


// This function updates the contents of a room by recalculating its capacity and other attributes
exports.updateRoom = async (req, res) => {

  let reqJson = JSON.parse(req.body);
  console.log(reqJson);
  
try {
  let room =new Room(0,reqJson.roomNumber, reqJson.roomName, reqJson.floorNumber,reqJson.roomArea,reqJson.deskArea,reqJson.numberDesks,reqJson.capacityPercentage);
  let roomData = {
    floorNumber:room.floorNumber,
    roomNumber:room.roomNumber,
    roomName:room.roomName,
    roomArea:room.roomArea,
    capacityPercentage:room.capacityPercentage,
    numberDesks: room.numberDesks,
    occupiedDesks:room.occupiedDesks,
    currentCapacity:room.currentCapacity,
    deskArea:room.deskArea,
    capacityOfPeopleForSixFtGrid:room.capacityOfPeopleForSixFtGrid,
    capacityOfPeopleForSixFtCircle:room.capacityOfPeopleForSixFtCircle
  }
  await database.editRoom(roomData);
  
  return res.status(200).send({
    message: 'room successfully updated',
    data: reqJson
  });
} catch (error) {
  console.log(error);
  return res.status(500).send(error);
}

};


//Deletes a single room with all its desks
exports.deleteRoom = async (req, res) => {
  try {
        let desks= await database.getDesks();

    let reqJson = JSON.parse(req.body);
    console.log(reqJson);

        desks.forEach(obj4 =>{
        if(obj4.roomNumber===reqJson.roomNumber)
        {
           database.deleteDesk(obj4.deskNumber);
        }
        else
        {
  
        }

        });

    let floor = await database.getFloor(reqJson.floorNumber);
    await database.removeRoom(reqJson.floorNumber,floor.numRooms)
    await database.deleteRoom(reqJson.roomNumber);

    // company-data summary
    let companyData = await reportingDatabase.getCompanyData(floor.companyId);
    await reportingDatabase.decreaseNumberOfRoomsCompanyData(floor.companyId, companyData.numberOfRooms);

      return res.status(200).send({
        message: 'Room successfully deleted',
      });
  } catch (error) {
    console.log(error);
    return res.status(500).send(error);
  }
};

//Deletes a floor with all its rooms and desks
exports.deleteFloor = async (req, res) => {
  try {
    let filteredList=[];
    let rooms = await database.getRooms();
    let numRoomsToDelete = 0;

    let reqJson = JSON.parse(req.body);
    console.log(reqJson);

    rooms.forEach(obj => {
      if(obj.floorNumber===reqJson.floorNumber)
      {
        filteredList.push(obj);
        numRoomsToDelete++;
      }
      else
      {

      }
    });

    let desks= await database.getDesks();
    filteredList.forEach(obj => {
      
        desks.forEach(obj4 =>{
        if(obj4.roomNumber===obj.roomNumber)
        {
          database.deleteDesk(obj4.deskNumber);
        }
        else
        {
  
        }

        });

    database.deleteRoom(obj.roomNumber);
    });
    let floorplan= await  database.getFloorPlan(reqJson.floorplanNumber);
    await database.removeFloor(reqJson.floorplanNumber,floorplan.numFloors);
    await database.deleteFloor(reqJson.floorNumber);

    // company-data summary
    let companyData = await reportingDatabase.getCompanyData(floorplan.companyId);
    await reportingDatabase.decreaseNumberOfFloorsCompanyData(floorplan.companyId, companyData.numberOfFloors);
    for (let i = 0; i < numRoomsToDelete; i++) {
      await reportingDatabase.decreaseNumberOfRoomsCompanyData(floorplan.companyId, parseInt(companyData.numberOfRooms) - i);
    }

      return res.status(200).send({
        message: 'Floor successfully deleted',
      });
  } catch (error) {
    console.log(error);
    return res.status(500).send(error);
  }
};


//Deletes a floorplan with all its floors and rooms and desks
exports.deleteFloorPlan = async (req, res) => {
  try {
    let filteredList=[];
    let floors = await database.getFloors();
    let numFloorsToDelete = 0;
    let numRoomsToDelete = 0;

    let reqJson = JSON.parse(req.body);
    console.log(reqJson);

    let floorplan = await database.getFloorPlan(reqJson.floorplanNumber);

    floors.forEach(obj => {
      if(obj.floorplanNumber===reqJson.floorplanNumber)
      {
        filteredList.push(obj);
        numFloorsToDelete++;
      }
      else
      {

      }
    });
    let desks= await database.getDesks();
    let rooms = await database.getRooms();
    filteredList.forEach(obj => {
      let filteredList2=[];
      rooms.forEach(obj2 => {
        if(obj2.floorNumber===obj.floorNumber)
        {
          filteredList2.push(obj2);
          numRoomsToDelete++;
        }
        else
        {
  
        }
      });
      
     
      filteredList2.forEach(obj3 => {
        
        desks.forEach(obj4 =>{
        if(obj4.roomNumber===obj3.roomNumber)
        {
           database.deleteDesk(obj4.deskNumber);
        }
        else
        {
  
        }

        });
        database.deleteRoom(obj3.roomNumber);
        });
    database.deleteFloor(obj.floorNumber);
    });

    await database.deleteFloorPlan(reqJson.floorplanNumber);

    // company-data summary
    let companyData = await reportingDatabase.getCompanyData(floorplan.companyId);
    await reportingDatabase.decreaseNumberOfFloorplansCompanyData(floorplan.companyId, companyData.numberOfFloorplans);
    for (let i = 0; i < numFloorsToDelete; i++) {
      await reportingDatabase.decreaseNumberOfFloorsCompanyData(floorplan.companyId, parseInt(companyData.numberOfFloors) - i);
    }
    for (let i = 0; i < numRoomsToDelete; i++) {
      await reportingDatabase.decreaseNumberOfRoomsCompanyData(floorplan.companyId, parseInt(companyData.numberOfRooms) - i);
    }

      return res.status(200).send({
        message: 'Floor Plan successfully deleted',
      });
  } catch (error) {
    console.log(error);
    return res.status(500).send(error);
  }
};

File: frontend/lib/controllers/floor_plan/floor_plan_controller.dart
// Floor plan controller
library controllers;

import 'package:http/http.dart' as http;
import 'dart:async';
import 'dart:convert';

import 'package:frontend/models/floor_plan/floor_plan.dart';
import 'package:frontend/models/floor_plan/floor.dart';
import 'package:frontend/models/floor_plan/room.dart';
import 'package:frontend/controllers/server_info.dart' as serverInfo;

List<FloorPlan> floorPlanDatabaseTable = [];
int numFloorPlans = 0;

List<Floor> floorDatabaseTable = [];
int numFloors = 0;

List<Room> roomDatabaseTable = [];
int numRooms = 0;

String server = serverInfo.getServer(); //server needs to be running on firebase

Future<bool> createFloorPlan(num numFloors, String adminId, String companyId, String imageBytes) async {
  String path = "/floorplan";
  String url = server + path;
  var request;

  try {
    request = http.Request('POST', Uri.parse(url));
    request.body = json.encode({
      "numFloors": numFloors,
      "adminId": adminId,
      "companyId": companyId,
      "base64String": imageBytes,
    });

    var response = await request.send();

    if (response.statusCode == 200) {
      print(await response.stream.bytesToString());

  return true;
  }
  } catch (error) {
  print(error);
  }

  return false;
}

Future<bool> createFloor(String floorPlanNumber, String adminId, String companyId) async {
  String path = "/floorplan/floor";
  String url = server + path;
  var request;

  try {
    request = http.Request('POST', Uri.parse(url));
    request.body = json.encode({
      "floorplanNumber": floorPlanNumber,
      "adminId": adminId,
      "companyId": companyId,
    });

    var response = await request.send();

    if (response.statusCode == 200) {
      print(await response.stream.bytesToString());

      return true;
    }
  } catch (error) {
    print(error);
  }

  return false;
}

Future<bool> createRoom(num currentNumRoomsInFloor, String floorNumber, String imageBytes) async {
  String path = "/floorplan/room";
  String url = server + path;
  var request;

  try {
    request = http.Request('POST', Uri.parse(url));
    request.body = json.encode({
      "currentNumberRoomInFloor": currentNumRoomsInFloor,
      "floorNumber": floorNumber,
      "roomArea": 0,
      "capacityPercentage": 0,
      "numberDesks": 0,
      "occupiedDesks": 0,
      "currentCapacity": 0,
      "deskArea": 0,
      "capacityOfPeopleForSixFtGrid": 0,
      "capacityOfPeopleForSixFtCircle": 0,
      "base64String": imageBytes,
    });

    print(request.body);

    var response = await request.send();

    if (response.statusCode == 200) {
      print(await response.stream.bytesToString());

      return true;
    }
  } catch (error) {
    print(error);
  }

  return false;
}

Future<List<FloorPlan>> getFloorPlans(String companyId) async {
  String path = '/floorplan/view';
  String url = server + path;
  var request;

  try {
    request = http.Request('POST', Uri.parse(url));
    request.body = json.encode({
      "companyId": companyId
    });

    var response = await request.send();

    if (response.statusCode == 200) {
      var jsonString = (await response.stream.bytesToString());
      var jsonMap = jsonDecode(jsonString);

      //Added these lines so that it doesn't just keep adding and adding to the list indefinitely everytime this function is called
      floorPlanDatabaseTable.clear();
      numFloorPlans = 0;

      for (var data in jsonMap["data"]) {
        var floorPlanData = FloorPlan.fromJson(data);
        floorPlanDatabaseTable.add(floorPlanData);
        numFloorPlans++;
      }

      return floorPlanDatabaseTable;
    }
  } catch (error) {
    print(error);
  }

  return null;
}

Future<List<Floor>> getFloors(String floorPlanNumber) async {
  String path = '/floorplan/floors';
  String url = server + path;
  var request;

  try {
    request = http.Request('POST', Uri.parse(url));
    request.body = json.encode({
      "floorplanNumber": floorPlanNumber
    });

    var response = await request.send();

    if (response.statusCode == 200) {
      var jsonString = (await response.stream.bytesToString());
      var jsonMap = jsonDecode(jsonString);

      //Added these lines so that it doesn't just keep adding and adding to the list indefinitely everytime this function is called
      floorDatabaseTable.clear();
      numFloors = 0;

      for (var data in jsonMap["data"]) {
        var floorData = Floor.fromJson(data);
        floorDatabaseTable.add(floorData);
        numFloors++;
      }

      return floorDatabaseTable;
    }
  } catch (error) {
    print(error);
  }

  return null;
}

Future<List<Room>> getRooms(String floorNumber) async {
  String path = '/floorplan/floors/rooms';
  String url = server + path;
  var request;

  try {
    request = http.Request('POST', Uri.parse(url));
    request.body = json.encode({
      "floorNumber": floorNumber
    });

    var response = await request.send();

    if (response.statusCode == 200) {
      var jsonString = (await response.stream.bytesToString());
      var jsonMap = jsonDecode(jsonString);

      //Added these lines so that it doesn't just keep adding and adding to the list indefinitely everytime this function is called
      roomDatabaseTable.clear();
      numRooms = 0;

      for (var data in jsonMap["data"]) {
        var roomData = Room.fromJson(data);
        roomDatabaseTable.add(roomData);
        numRooms++;
      }

      return roomDatabaseTable;
    }
  } catch (error) {
    print(error);
  }

  return null;
}

Future<bool> updateRoom(String floorNumber, String roomNumber, String roomName, num roomArea,
    num numberOfDesks, num deskArea, num capacityPercentage, String imageBytes) async {
  String path = '/floorplan/room';
  String url = server + path;
  var request;

  try {
    request = http.Request('PUT', Uri.parse(url));
    request.body = json.encode({
      "floorNumber": floorNumber,
      "roomNumber": roomNumber,
      "roomName": roomName,
      "roomArea": roomArea,
      "numberDesks": numberOfDesks,
      "deskArea": deskArea,
      "capacityPercentage": capacityPercentage,
      "base64String": imageBytes,
    });

    var response = await request.send();

    if (response.statusCode == 200) {
      print(await response.stream.bytesToString());

      return true;
    }
  } catch (error) {
    print(error);
  }

  return false;
}

Future<bool> deleteFloorPlan(String floorPlanNumber) async {
  String path = '/floorplan';
  String url = server + path;

  var request = http.Request('DELETE', Uri.parse(url));
  request.body = json.encode({"floorplanNumber": floorPlanNumber});

  var response = await request.send();

  if (response.statusCode == 200) {
    print(await response.stream.bytesToString());

    for (int i = 0; i < floorPlanDatabaseTable.length; i++) {
      if (floorPlanDatabaseTable[i].floorPlanNumber == floorPlanNumber) {
        floorPlanDatabaseTable.removeAt(i);
        numFloorPlans--;
      }
    }

    return true;
  }

  //Double check to make sure it isn't still being stored internally
  for (int i = 0; i < numFloorPlans; i++) {
    if (floorPlanDatabaseTable[i].floorPlanNumber == floorPlanNumber) {
      floorPlanDatabaseTable.removeAt(i);
      numFloorPlans--;
    }
  }

  return false;
}

Future<bool> deleteFloor(String floorPlanNumber, String floorNumber) async {
  String path = '/floorplan/floor';
  String url = server + path;

  var request = http.Request('DELETE', Uri.parse(url));
  request.body = json.encode({
    "floorplanNumber": floorPlanNumber,
    "floorNumber": floorNumber,
  });

  var response = await request.send();

  if (response.statusCode == 200) {
    print(await response.stream.bytesToString());

    for (int i = 0; i < floorDatabaseTable.length; i++) {
      if (floorDatabaseTable[i].floorNumber == floorNumber) {
        floorDatabaseTable.removeAt(i);
        numFloors--;
      }
    }

    return true;
  }

  //Double check to make sure it isn't still being stored internally
  for (int i = 0; i < numFloors; i++) {
    if (floorDatabaseTable[i].floorNumber == floorNumber) {
      floorDatabaseTable.removeAt(i);
      numFloors--;
    }
  }

  return false;
}

Future<bool> deleteRoom(String floorNumber, String roomNumber) async {
  String path = '/floorplan/room';
  String url = server + path;

  var request = http.Request('DELETE', Uri.parse(url));
  request.body = json.encode({
    "floorNumber": floorNumber,
    "roomNumber": roomNumber,
  });

  var response = await request.send();

  if (response.statusCode == 200) {
    print(await response.stream.bytesToString());

    for (int i = 0; i < roomDatabaseTable.length; i++) {
      if (roomDatabaseTable[i].roomNumber == roomNumber) {
        roomDatabaseTable.removeAt(i);
      }
    }

    return true;
  }

  return false;
}